window_management/compass_control.py

```diff
@@ -0,0 +1,2052 @@
+                print(f"get_component_dimensions: y steps={y_steps}")
+        else:
+            if direction_count == 1:    # horizontal or vertical
+                if direction["left"] or direction["right"]:
+                    delta_width = distance
+                elif direction["up"] or direction["down"]:
+                    delta_height = distance
+            else:  # diagonal
+                diagonal_length = self.get_diagonal_length(rect)
+                ratio = distance / diagonal_length
+                delta_width = rect.width * ratio
+                delta_height = rect.height * ratio
+
+        if self.testing:
+            print(f"get_component_dimensions: returning {delta_width}, {delta_height}")
+
+        return round(delta_width), round(delta_height)
+
+    def get_component_dimensions_by_percent(self, rect: ui.Rect, rect_id: int, parent_rect: ui.Rect,
+                                                percent: float, direction: Direction, operation: str) -> Tuple[int, int]:
+        """Return horizontal and vertical distances corresponding to the given percentage of the diagonal of the given rectangle"""
+        if self.testing:
+            print(f'_get_component_dimensions_by_percent: {percent=}')
+
+        direction_count = sum(direction.values())
+        if operation == 'move':
+            if direction_count == 0:    # in
+                rect, *unused = self.get_center_to_center_rect(rect, rect_id, parent_rect)
+            elif direction_count == 4:
+                rect, *unused = self.get_center_to_intercept_rect(rect, rect_id, parent_rect)
+
+        if direction_count  == 1:    # horizontal or vertical
+            if direction["left"] or direction["right"]:
+                distance = rect.width * (percent/100)
+            elif direction["up"] or direction["down"]:
+                distance =  rect.height * (percent/100)
+        else:  # diagonal
+            diagonal_length = self.get_diagonal_length(rect)
+            distance = diagonal_length * (percent/100)
+
+        return self.get_component_dimensions(rect, rect_id, parent_rect, distance, direction, operation)
+
+    # Bresenham line code, from
+    #       https://github.com/encukou/bresenham/blob/master/bresenham.py
+    def bresenham(self, x0: int, y0: int, x1: int, y1: int) -> Tuple[int, int]:
+        """Yield integer coordinates on the line from (x0, y0) to (x1, y1).
+
+        Input coordinates should be integers.
+
+        The result will contain both the start and the end point.
+        """
+        dx = x1 - x0
+        dy = y1 - y0
+
+        xsign = 1 if dx > 0 else -1
+        ysign = 1 if dy > 0 else -1
+
+        dx = abs(dx)
+        dy = abs(dy)
+
+        if dx > dy:
+            xx, xy, yx, yy = xsign, 0, 0, ysign
+        else:
+            dx, dy = dy, dx
+            xx, xy, yx, yy = 0, ysign, xsign, 0
+
+        D = 2*dy - dx
+        y = 0
+
+        for x in range(dx + 1):
```

Comment:
rects will contains floats on Mac, so this will presently crash here when using the continuous commands. May be easiest just to cast dx/dy

Reply:
That was fast - thanks! I think the change I just pushed covers it, can you check?